### PR TITLE
gh-111495: Add tests for PyFloat C API

### PR DESCRIPTION
**Pull request by skirpichev**
* Issue: gh-111495

### CODE REVIEW

**Comment by skirpichev** [comment]
Some C API tests for floats are in ``Lib/test/test_float.py`` historically.  An example: https://github.com/python/cpython/blob/ed587be0d0383f2925bf7650e8ccf1bf3adc44f9/Lib/test/test_float.py#L1518-L1569

In this PR this left intact, but maybe we should move them.

File: Lib/test/test_capi/test_float.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
+import array
+import math
+import sys
+import unittest
+import warnings
+
+from test.test_capi.test_getargs import (Float, FloatSubclass, FloatSubclass2,
```
Comment (sobolevn) [inline Lib/test/test_capi/test_float.py:6]: It is better to not import from one test in other tests, I propose adding a `utils` module for common types / helpers.

File: Lib/test/test_capi/test_float.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+        # CRASHES checkexact(NULL)
+
+    def test_fromstring(self):
+        # Test PyFloat_FromString()
+        fromstring = _testcapi.float_fromstring
+
+        self.assertEqual(fromstring("3.14"), 3.14)
+        self.assertEqual(fromstring(b"3.14"), 3.14)
+        self.assertEqual(fromstring(bytearray(b"3.14")), 3.14)
```
Comment (sobolevn) [inline Lib/test/test_capi/test_float.py]: What about `memoryview`?

File: Lib/test/test_capi/test_float.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        self.assertEqual(fromstring(array.array('b', b'3.14')), 3.14)
+        self.assertEqual(fromstring("10."), 10.)
+        self.assertEqual(fromstring("-1e10"), -1e10)
+        self.assertEqual(fromstring("3.14e+42"), 3.14e+42)
+        self.assertEqual(fromstring("0e0"), 0.0)
```
Comment (sobolevn) [inline Lib/test/test_capi/test_float.py]: What about `-inf` and `inf` and `NaN`?

File: Lib/test/test_capi/test_float.py (hunk 2 above)
Comment (skirpichev) [reply Lib/test/test_capi/test_float.py]: That seems redundant: array was used above to test conversion from objects that support buffer protocol.

File: Lib/test/test_capi/test_float.py (hunk 3 above)
Comment (skirpichev) [reply Lib/test/test_capi/test_float.py]: In fact, I'm thinking about removing most current examples with strings.  That could go to PyOs_* tests (``PyOS_string_to_double()``).

Instead, here we should add more tests with underscores to cover ``_Py_string_to_number_with_underscores()`` helper.

File: Lib/test/test_capi/test_float.py (hunk 1 above)
Comment (skirpichev) [reply Lib/test/test_capi/test_float.py:6]: Yes, I was already thinking about such refactoring.  See also #110956.

Does it make sense to do here (at least for affected support classes) or it belongs to a separate pr?

Comment (sobolevn) [reply Lib/test/test_capi/test_float.py:6]: I think that this should be done separately.

File: Lib/test/test_capi/test_float.py (hunk 3 above)
Comment (skirpichev) [reply Lib/test/test_capi/test_float.py]: In fact, as parsing with whitespaces/underscores tested in test/test_float.py - this was removed as well.

File: Lib/test/test_capi/test_float.py (hunk 1 above)
Comment (skirpichev) [reply Lib/test/test_capi/test_float.py:6]: https://github.com/python/cpython/issues/111673

File: Lib/test/test_capi/test_float.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+    def test_asdouble(self):
+        # Test PyFloat_AsDouble()
+        asdouble = _testcapi.float_asdouble
+
+        class BadFloat3:
+            def __float__(self):
+                raise RuntimeError
+
+        self.assertEqual(asdouble(3.14), 3.14)
```
Comment (sobolevn) [inline Lib/test/test_capi/test_float.py]: I think that we need to test `-1.` here, because it is a special error value as well.

**Review by sobolevn — COMMENTED** [review]
Just one suggestion :)
All other things LGTM.

File: Lib/test/test_capi/test_float.py (hunk 4 above)
Comment (skirpichev) [inline Lib/test/test_capi/test_float.py]:
```suggestion
        self.assertEqual(asdouble(3.14), 3.14)
        self.assertEqual(asdouble(-1), -1.0)
```

Comment (skirpichev) [reply Lib/test/test_capi/test_float.py]: Ok, lets do it.

**Comment by skirpichev** [comment]
Converted to a draft per review #111591

File: Lib/test/test_capi/test_float.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        self.assertFalse(checkexact(object()))
+
+        # CRASHES checkexact(NULL)
+
+    def test_fromstring(self):
+        # Test PyFloat_FromString()
+        fromstring = _testcapi.float_fromstring
+
+        self.assertEqual(fromstring("3.14"), 3.14)
+        self.assertEqual(fromstring(b"3.14"), 3.14)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_float.py]: Add tests for embedded null.
`'3.14\0'`
`b'3.14\0'`
if they raise or
`'3.14\0z'`
`b'3.14\0z'`
if they silently truncate.

File: Lib/test/test_capi/test_float.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        self.assertEqual(fromstring(bytearray(b"3.14")), 3.14)
+        self.assertEqual(fromstring(array.array('b', b'3.14')), 3.14)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_float.py]: Add tests for memoryview:
`memoryview(b'3.14')`
`memoryview(b'3.141')[:-1]`  # not null-terminated
`memoryview(b'3.14\0')[:-1]`  (if the above fails)
`memoryview(b'3.14')[::2]`  # non-continuous

File: Lib/test/test_capi/test_float.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    def test_asdouble(self):
+        # Test PyFloat_AsDouble()
+        asdouble = _testcapi.float_asdouble
+
+        class BadFloat3:
+            def __float__(self):
+                raise RuntimeError
+
+        self.assertEqual(asdouble(3.14), 3.14)
+        self.assertEqual(asdouble(-1), -1.0)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_float.py:77]: Test also positive integer and -1.0.

Test also with too large integer like 2**1000.

File: Lib/test/test_capi/test_float.py (hunk 6 above)
Comment (skirpichev) [reply Lib/test/test_capi/test_float.py]: This was a test for buffer protocol.  Maybe I should replace array example?

Comment (serhiy-storchaka) [reply Lib/test/test_capi/test_float.py]: Yes, memoryview example can replace array example.

But it allow to test more interesting cases.

File: Modules/_testcapi/float.c [hunk 8]
```diff
@@ -2,9 +2,61 @@
...
+
+static PyObject *
+float_fromstring(PyObject *Py_UNUSED(module), PyObject *obj)
+{
+    NULLABLE(obj);
+    return PyFloat_FromString(obj);
+}
+
+static PyObject *
+float_asdouble(PyObject *Py_UNUSED(module), PyObject *obj)
```
Comment (serhiy-storchaka) [inline Modules/_testcapi/float.c]: Wait, this wrapper is for testing `PyFloat_AsDouble`, but it does not contain `PyFloat_AsDouble`!

Comment (skirpichev) [reply Modules/_testcapi/float.c]: Good catch!  This was expected to be the new wrapper float_fromdouble().

File: Lib/test/test_capi/test_float.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+_testcapi = import_helper.import_module('_testcapi')
+
+NULL = None
+
+
+class CAPIFloatTest(unittest.TestCase):
+    def assertEqualWithSign(self, actual, expected):
+        self.assertEqual(actual, expected)
+        self.assertEqual(math.copysign(1, actual), math.copysign(1, expected))
+
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_float.py]: Not used.
```suggestion
```

File: Lib/test/test_capi/test_float.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    def test_check(self):
+        # Test PyFloat_Check()
+        check = _testcapi.float_check
+
+        self.assertTrue(check(4.25))
+        self.assertTrue(check(FloatSubclass(4.25)))
+        self.assertFalse(check(Float()))
+        self.assertFalse(check(3))
+        self.assertFalse(check([]))
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_float.py]: Lists are not float-like in any way.
```suggestion
```

File: Lib/test/test_capi/test_float.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+    def test_checkexact(self):
+        # Test PyFloat_CheckExact()
+        checkexact = _testcapi.float_checkexact
+
+        self.assertTrue(checkexact(4.25))
+        self.assertFalse(checkexact(FloatSubclass(4.25)))
+        self.assertFalse(checkexact(Float()))
+        self.assertFalse(checkexact(3))
+        self.assertFalse(checkexact([]))
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_float.py]:
```suggestion
```

File: Lib/test/test_capi/test_float.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+        # Test PyFloat_AsDouble()
+        asdouble = _testcapi.float_asdouble
+
+        class BadFloat3:
+            def __float__(self):
+                raise RuntimeError
+
+        self.assertEqual(asdouble(4.25), 4.25)
+        self.assertEqual(asdouble(-1), -1.0)
+        self.assertEqual(asdouble(42), 42.0)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_float.py]:
```suggestion
        self.assertEqual(asdouble(-1.0), -1.0)
        self.assertEqual(asdouble(42), 42.0)
        self.assertEqual(asdouble(-1), -1.0)
```

File: Modules/_testcapi/float.c [hunk 13]
```diff
@@ -2,9 +2,75 @@
...
+    NULLABLE(obj);
+    d = PyFloat_AsDouble(obj);
+    if (d == -1. && PyErr_Occurred()) {
+        return NULL;
+    }
+
+    return PyFloat_FromDouble(d);
+}
+
+static PyObject*
```
Comment (serhiy-storchaka) [inline Modules/_testcapi/float.c]:
```suggestion
static PyObject *
```

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

Thank you @skirpichev for your contribution and for your fast response.

**Comment by skirpichev** [comment]
@serhiy-storchaka, thank you for review.

**Pull request merged**